### PR TITLE
refacto(backend): poetry package manager and chat route refactoring 

### PR DESCRIPTION
**Pull request by AmineDiro**
# Description
- Added package manager
- Added precommit checks
- Rewrote dependency injection of Services and Repositories
- Integrate async SQL alchemy engine
- Migrate Chat  repository to SQLModel 
- Migrated ChatHistory repository to SQLModel
- User SQLModel
- Unit test methodology with db rollback
- Unit tests ChatRepository
- Test ChatService get_history
- Brain entity SQL Model
- Promp SQLModel
- Rewrite chat/{chat_id}/question route
- updated docker files and docker compose in dev and production

Added `quivr_core` subpackages:
- Refactored KnowledgebrainQa
- Added Rag service to interface with non-rag dependencies

### CODE REVIEW

**Review by StanGirard — COMMENTED** [review]
Thanks :) I'll keep reviewing

File: backend/backend/celery_config.py [hunk 1]
```diff
@@ -39,6 +39,3 @@
     )
 else:
     raise ValueError(f"Unsupported broker URL: {CELERY_BROKER_URL}")
-
-
-celery.autodiscover_tasks(["modules.sync", "modules", "middlewares", "packages"])
```
Comment (StanGirard) [inline backend/backend/celery_config.py:44]: @AmineDiro t'es sur de toi ? Ca permert de charger les celery et les packages dont ils ont besoin car celery ne fait pas d'autoimport ?

Comment (AmineDiro) [reply backend/backend/celery_config.py:44]: On n'a pas besoin parce qu'on a le package manager maintenant 🙂 les tasks sont dans un modules donc pas besoin de pointer sur les fichiers

File: backend/supabase/migrations/20240606083114_index-creation.sql
Comment (StanGirard) [inline backend/supabase/migrations/20240606083114_index-creation.sql:1]: index créé

File: backend/backend/celery_worker.py [hunk 2]
```diff
@@ -264,28 +254,20 @@ def check_if_is_premium_user():
                     "product"
                 ]
             ]
-            if len(matching_product_settings) > 0:
```
Comment (StanGirard) [inline backend/backend/celery_worker.py:267]: check ca

**Comment by StanGirard** [comment]
Le sync assistant n'est pas loadé

File: backend/backend/celery_worker.py (hunk 2 above)
Comment (AmineDiro) [reply backend/backend/celery_worker.py:267]: fixed rebase issue on this file

File: backend/supabase/migrations/20240606083114_index-creation.sql
Comment (AmineDiro) [reply backend/supabase/migrations/20240606083114_index-creation.sql:1]: Fixed in [3956668](https://github.com/QuivrHQ/quivr/pull/2684/commits/39566680317b9645698bcebc1367a86a3408fbcc)

File: backend/backend/packages/quivr_core/brain_factory.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+            else None
+        )
+
+    def get_or_create_brain(self, brain_id: UUID | None, user_id: UUID) -> BrainEntity:
+        brain = None
+        if brain_id is not None:
+            brain = self.brain_service.get_brain_details(brain_id, user_id)
+
+        # TODO: Create if doesn't exist
+        assert brain
```
[the remark this answers is not in the document]
Comment (AmineDiro) [reply backend/backend/packages/quivr_core/brain_factory.py:98]: You're right. From what I can gather, the only Rag type is the `BrainQA`  🤔 ?

**Comment by StanGirard** [comment]
Sounds good to me if you fix the conflict ;) 

Also we need to prioritize if no brains selected ASAP

**Pull request merged**